crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -1364,6 +1401,58 @@ def upgrade_compute_nodes(virt)
         compute_nodes.each do |n|
           upgrade_compute_node(controller, n)
         end
+      end
+
+      # Non-disruptive way of upgrading more compute nodes at once.
+      # Does live-evacuation of each compute node first, then starts parallel upgrade.
+      # Argument is list of node names.
+      def non_disruptive_upgrade_compute_nodes(names)
+        compute_nodes = names.map do |name|
+          ::Node.find_node_by_name_or_alias(name)
+        end
+
+        controller = fetch_nova_controller
+
+        # If there's a compute node which we already started to upgrade,
+        # (and the upgrade process was restarted due to the failure)
+        # continue with that one.
+        compute_nodes.sort! { |n| n.upgrading? ? -1 : 1 }
+
+        while compute_nodes.any?
+          # 1. Evacuate as many as possible, create a subset of nodes without instances.
+          nodes_to_upgrade = []
+          already_upgrading = false
+          compute_nodes.each do |node|
+            # if nodes are already in upgrading state, we can skip the live-migration part
+            already_upgrading = true if node.upgrading?
+            break if already_upgrading && !node.upgrading?
+            hostname = node[:hostname]
+            begin
+              live_evacuate_compute_node(controller, hostname) unless already_upgrading
+              nodes_to_upgrade << node
+            rescue StandardError => e
+              # We could safely ignore the error when the execution failed with a timeout,
+              # raise it otherwise.
+              failed_file = "/var/lib/crowbar/upgrade/crowbar-evacuate-host-failed"
+              raise e if controller.file_exist? failed_file
+              Rails.logger.info("attempt to live evacuate #{hostname} took too long, exiting loop")
+              # We need to enable nova-service on the failed node again so its compute powers could
+              # be used when upgrading next bunch of compute nodes
+              enable_compute_service(controller, node, true)
```

Comment:
Is this a good idea? We don't exactly know the evacuation failed on this node. Is it smart to put it back into the pool?

Reply:
That's indeed problematic, I know. However I'm trying to rule out the case of an error on lines 1436-1437. If script itself found an error that it should have created the failed file.

If there is no failed file, than the script was killed by timeout which (likely, I know) means that live evacuation failed because there was no other host available.

Again, I know I cannot be 100% sure about this. But _not_ putting the node back into the pool could bring a risk of having increased number of nodes with nova-compute service disabled which would reduce a chance of live-migration in the next iteration.